pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/repositories/impl/RedisPassOwnerRepository.kt

```diff
@@ -0,0 +1,98 @@
+) : PassOwnerRepository by mongoPassOwnerRepository {
+
+    override fun findById(passOwnerId: String): Mono<MongoPassOwner> {
+        val key = passOwnerKey(passOwnerId)
+        return reactiveRedisTemplate.opsForValue().get(key)
+            .handle { item, sink ->
+                if (item.isEmpty()) {
+                    sink.error(PassOwnerNotFoundException("Could not find pass owner by id $passOwnerId"))
+                } else {
+                    sink.next(item)
+                }
+            }
+            .map { objectMapper.readValue<MongoPassOwner>(it) }
+            .switchIfEmpty {
+                mongoPassOwnerRepository.findById(passOwnerId)
+                    .flatMap(::savePassOwnerToRedis)
+                    .switchIfEmpty {
+                        reactiveRedisTemplate.opsForValue()
+                            .set(key, byteArrayOf(), Duration.ofMinutes(redisExpirationTimeoutInMinutes))
+                            .then(Mono.empty())
+                    }
+            }
+            .onErrorResume(::isRedisOrSocketException) { mongoPassOwnerRepository.findById(passOwnerId) }
+    }
+
+    override fun insert(newMongoPassOwner: MongoPassOwner): Mono<MongoPassOwner> {
+        return mongoPassOwnerRepository.insert(newMongoPassOwner).flatMap { created ->
+            savePassOwnerToRedis(created)
+                .onErrorResume(::isRedisOrSocketException) {
+                    log.warn("Redis failed to cache created user.", it)
+                    Mono.empty()
+                }.thenReturn(created)
+        }
+    }
+
+    override fun deleteById(passOwnerId: String): Mono<Unit> {
+        val key = passOwnerKey(passOwnerId)
+        return mongoPassOwnerRepository.deleteById(passOwnerId)
+            .flatMap {
+                reactiveRedisTemplate.opsForValue().delete(key)
+                    .onErrorResume(::isRedisOrSocketException) {
+                        log.warn("Redis failed to remove user from cache.", it)
+                        Mono.empty()
+                    }.thenReturn(Unit)
+            }
+    }
+
+    override fun save(newPassOwner: MongoPassOwner): Mono<MongoPassOwner> {
+        return mongoPassOwnerRepository.save(newPassOwner).flatMap { saved ->
+            savePassOwnerToRedis(saved).onErrorResume(::isRedisOrSocketException) {
+                log.warn("Redis failed to cache created user.", it)
+                Mono.empty()
+            }.thenReturn(saved)
+        }
+    }
+
+    private fun savePassOwnerToRedis(passOwner: MongoPassOwner): Mono<MongoPassOwner> {
+        val key = passOwnerKey(passOwner.id.toString())
+        val byteArray = objectMapper.writeValueAsBytes(passOwner)
+        return reactiveRedisTemplate.opsForValue().set(
+            key,
+            byteArray,
+            Duration.ofMinutes(redisExpirationTimeoutInMinutes)
+        ).thenReturn(passOwner)
+    }
+
+    companion object {
+        private const val KEY_PREFIX = "key-"
```

Comment:
what if you save more entities together with the pass owner in redis? even If no collision by id? how will you differentiate entries from each other?